monai/transforms/post/array.py

```diff
@@ -289,6 +292,144 @@ def __call__(self, img: torch.Tensor) -> torch.Tensor:
+              |               \  |  /                 |  <- hop 2
+        [ ]--[x]--[ ]      [ ]--[x]--[ ]        [x]--[ ]
+              |               /  |  \             hop 1
+             [ ]           [ ]  [ ]  [ ]
+
+    It is possible to define for which labels the hole filling should be applied.
+    The input image is assumed to be a PyTorch Tensor or numpy array
+    with shape [batch_size, 1, spatial_dim1[, spatial_dim2, ...]] and the values correspond to expected labels.
+
+    Note:
+
+        The label 0 will be treated as background and the enclosed holes will be set to the neighboring class label.
+
+        The performance of this method heavily depends on the number of labels.
+        It is a bit faster if the list of `applied_labels` is provided.
+        Limiting the number of `applied_labels` results in a big decrease in processing time.
+
+    For example:
+
+        Use FillHoles with default parameters::
+
+            [1, 1, 1, 2, 2, 2, 3, 3]         [1, 1, 1, 2, 2, 2, 3, 3]
+            [1, 0, 1, 2, 0, 0, 3, 0]    =>   [1, 1 ,1, 2, 0, 0, 3, 0]
+            [1, 1, 1, 2, 2, 2, 3, 3]         [1, 1, 1, 2, 2, 2, 3, 3]
+
+        The hole in label 1 is fully enclosed and therefore filled with label 1.
+        The background label near label 2 and 3 is not fully enclosed and therefore not filled.
+    """
+
+    def __init__(
+        self, applied_labels: Optional[Union[Iterable[int], int]] = None, connectivity: Optional[int] = None
+    ) -> None:
+        """
+        Initialize the connectivity and limit the labels for which holes are filled.
+
+        Args:
+            applied_labels (Optional[Union[Iterable[int], int]], optional): Labels for which to fill holes. Defaults to None,
+                that is filling holes for all labels.
+            connectivity (int, optional): Maximum number of orthogonal hops to consider a pixel/voxel as a neighbor.
+                Accepted values are ranging from  1 to input.ndim. Defaults to a full
+                connectivity of ``input.ndim``.
+        """
+        super().__init__()
+        self.applied_labels = ensure_tuple(applied_labels) if applied_labels else None
+        self.connectivity = connectivity
+
+    def __call__(self, img: NdarrayTensor) -> NdarrayTensor:
+        """
+        Fill the holes in the provided image.
+
+        Note:
+            The value 0 is assumed as background label.
+
+        Args:
+            img (NdarrayTensor): Pytorch Tensor or numpy array
+            of shape [batch_size, num_channel, spatial_dim1[, spatial_dim2, ...]].
+
+        Raises:
+            NotImplementedError: The provided image was not a Pytorch Tensor or numpy array.
+
+        Returns:
+            Union[np.ndarray, torch.Tensor]: Pytorch Tensor or numpy array
+            of shape [batch_size, num_channel, spatial_dim1[, spatial_dim2, ...]].
+        """
+        if isinstance(img, np.ndarray):
+            channel_axis = 1
```

Comment:
Most of MONAI transforms work with `channel-first` data, so it should be `channel_axis = 0`, and you may need to also update the unit tests.

Thanks.

Reply:
This is channel first. Dimension 0 is the batch dimension.

Reply:
Hi @Spenhouet ,

The transform should work with single channel-first image, without batch dim.
You can refer to the `KeepLargestConnectedComponent`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/post/array.py#L257.

Thanks.

Reply:
You are right. I'm a bit confused since I remember having copy pasted the shape definition including the batch part from somewhere but I can't find it in the doc anywhere. Was this in a previous version or did I mix something up? What ever. I will fix it. Thanks for the hint.

Reply:
Just sad since I invested at least 30 min. in making the code work with batches.

Reply:
Hi @Spenhouet ,

Thanks for your quick update.
Yes, we used batch-first data for postprocessing transforms before, but changed the behavior since MONAI v0.6.
You can take a look at this doc:
https://github.com/Project-MONAI/MONAI/wiki/v0.5-to-v0.6-migration-guide#decollate-batch-first-tensor-to-list-of-channel-first-tensors
Now we decollate the batch to a list, then execute postprocessing transforms.

Thanks.

Reply:
I could leave the logic in to make it compatible with one-hot encoded data. Only the `applied_labels` filter wouldn't work then. It would always fill holes for all labels. It is valid to just throw a `NotImplementedError` in case someone runs this with one-hot and supplies the `applied_labels`? For non one-hot I would perform the squeeze then.

Reply:
I might actually have a solution to also make the `applied_labels` work for one-hot encoded data.
But I'm confused again. I can swear I did see a transform before that handled both cases - single channel and one-hot encoded data. Where did that go?

Reply:
Hi @Spenhouet ,

I think you can refer to the `KeepLargestConnectedComponent`, it can handle One-Hot:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/post/array.py#L273

Thanks.

Reply:
I believe there is a different one-hot handling needed here. But you will see it when I implemented it :)